### PR TITLE
EFS-137 Retry manager

### PR DESCRIPTION
**Pull request by gcatlin**
[EFS-137 Retry manager](https://spreedly.atlassian.net/browse/EFS-137)

### CODE REVIEW

**Comment by gcatlin** [comment]
The approach taken in this PR might be a horrible abuse of Erlang, but it's relatively simple and seems to work fairly well.  It would probably be more robust to create a separate Elixir OTP that runs on the Riak node like Commitlog was designed to do, but maybe the approach in the PR will be sufficient?

Here's how it work:
1. PCH attempts to send to Commitlog. If it succeeds, it sends a `{result, success}` message to the registered `retry_manager` process (which is `spawn`ed the first time it is needed).
2. If it failed, PCH sleeps then asks `retry_manager` if retries are enabled. If they are, it proceeds w/ retries using exponential back-off.
3. If all retries fail, PCH sends a `{result, failure}` message to the `retry_manager`, which will disable retries.
4. Retries will be disabled until PCH is able to send a message to Commitlog, so PCH will fail fast if Commitlog is down for more than a few seconds, but as soon as Commitlog comes back up and PCH can successfully communicate with it, retries will be enabled.

Thoughts?

**Comment by gcatlin** [comment]
One issue with the current implementation is that nothing is handling when `retry_manager` dies. It's not clear to me what the right solution is. Should I be using `spawn_monitor` instead? Will that prevent the parent process from being affected by the termination of the `retry_manager` process? And/or should `process_flag(trap_exit, true)` be used to convert an exit signal to an exit message?

**Comment by dtykocki** [comment]
@gcatlin - I'm trying to keep all the wires straight here.. does this PR replace #18?

**Comment by gcatlin** [comment]
@dtykocki, no it builds on it. I created a separate PR because you indicated that the original PR was hard to follow so I didn't want to make it worse.

The original PR (#18) is basically "PCH + Commitlog retries". This PR adds in a retry manager that provides the circuit-breaker-like functionality mentioned in #18.

**Comment by objectuser** [comment]
@gcatlin Heady stuff! 👍 

I think I need to reread some parts of the OTP book to remind myself about native messaging here.

File: src/postcommit_hook.erl [hunk 1]
```diff
@@ -66,23 +62,87 @@ sync_to_commitlog(Action, Bucket, Key, Value) ->
...
+send_to_retry_manager(Message) ->
+    ensure_retry_manager_is_registered(),
+    retry_manager ! Message.
+
+ensure_retry_manager_is_registered() ->
+    case whereis(retry_manager) of
+        undefined ->
+            Pid = spawn(fun() -> start_retry_manager() end),
+            try register(retry_manager, Pid)
+            catch _:badarg -> ensure_retry_manager_is_registered()
```
Comment (objectuser) [inline src/postcommit_hook.erl]: Looks like the retry manager is registered (VM-wide) under `retry_manager`. If I read your comments [here](https://github.com/spreedly/riak-postcommit-hook/pull/18/files#r301736639) correctly, since each PCH invocation is getting its own process, which means each PCH is executing concurrently, there's a race condition there right? That `badarg` exception [would be thrown](http://erlang.org/doc/man/erlang.html#register-2) and there would be an infinite loop.

Makes me wonder if there's a "startup hook" that can be used to register the process.

File: src/postcommit_hook.erl [hunk 2]
```diff
@@ -66,23 +62,87 @@ sync_to_commitlog(Action, Bucket, Key, Value) ->
...
+    end.
+
+send_to_retry_manager(Message) ->
+    ensure_retry_manager_is_registered(),
+    retry_manager ! Message.
+
+ensure_retry_manager_is_registered() ->
+    case whereis(retry_manager) of
+        undefined ->
+            Pid = spawn(fun() -> start_retry_manager() end),
```
Comment (dtykocki) [inline src/postcommit_hook.erl]: An issue with using `spawn` is that we won't know when a spawned process crashes or ends normally. If you use `spawn_monitor`, when the process goes down (with either a normal exit or a crash), you'll receive a message of the form `{'DOWN', MonitorReference, process, Pid, Reason}`

**Comment by objectuser** [comment]
I think we're at the point where we need some unit tests. 😄  The logic is sophisticated enough I think where that would be helpful.

Thoughts?

**Comment by objectuser** [comment]
The logic around the retry manager process and its usage seems solid to me. I think the questions currently are:

- How do we effectively start it and register it with the VM?
- How should we manage it with respect to crashes and monitoring?

I wonder if in the latter case, we shouldn't even worry about it. If it crashes, the consequence is we might release the flood gates to Riak. But in that case, we'd leverage it again. It might be enough to have some good logging around it (which might require `spawn_monitor` 😆 ).
